File: LeetCode/Python/Easy/1469-Find-All-The-Lonely-Nodes/solution.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getLonelyNodes(self, root: Optional[TreeNode]) -> List[int]:
        result=list()

        queue=deque()

        queue.append(root)

        while queue:
            q_size=len(queue)

            for iteration in range(0, q_size):
                parent=queue.popleft()

                children=0

                if parent.left and parent.right==None:
                    result.append(parent.left.val)

                    queue.append(parent.left)
                elif parent.left==None and parent.right:
                    result.append(parent.right.val)

                    queue.append(parent.right)
                elif parent.left and parent.right:
                    queue.append(parent.left)

                    queue.append(parent.right)

        return result

File: LeetCode/Rust/Easy/1469-Find-All-The-Lonely-Nodes/solution.rs
use std::rc::Rc;
use std::cell::{Ref, RefCell};
use std::collections::{VecDeque};
impl Solution
{
    pub fn get_lonely_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32>
    {
        let mut result: Vec<i32>=Vec::new();

        let mut queue: VecDeque<Option<Rc<RefCell<TreeNode>>>>=VecDeque::new();

        queue.push_back(root.clone());

        while let Some(parent)=queue.pop_back()
        {
            let parent_rc: &Rc<RefCell<TreeNode>>=parent.as_ref().unwrap();

            let parent_node: Ref<'_, TreeNode>=parent_rc.borrow();

            if parent_node.left.is_some() && parent_node.right.is_none()
            {
                result.push(parent_node.left.as_ref().unwrap().borrow().val);

                queue.push_back(parent_node.left.clone());
            }
            else if parent_node.left.is_none() && parent_node.right.is_some()
            {
                result.push(parent_node.right.as_ref().unwrap().borrow().val);

                queue.push_back(parent_node.right.clone());
            }
            else if parent_node.left.is_some() && parent_node.right.is_some()
            {
                queue.push_back(parent_node.left.clone());

                queue.push_back(parent_node.right.clone());
            }
        }

        result
    }
}

impl Solution
{
    pub fn get_lonely_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32>
    {
        let mut result: Vec<i32>=Vec::new();

        let mut queue: VecDeque<Option<Rc<RefCell<TreeNode>>>>=VecDeque::new();

        queue.push_back(root);

        while let Some(parent)=queue.pop_back()
        {
            let parent_rc: &Rc<RefCell<TreeNode>>=parent.as_ref().unwrap();

            let parent_node: Ref<'_, TreeNode>=parent_rc.borrow();

            match (parent_node.left.clone(), parent_node.right.clone())
            {
                (Some(left_child), None) =>
                {
                    result.push(left_child.borrow().val);

                    queue.push_back(Some(left_child));
                },
                (None, Some(right_child)) =>
                {
                    result.push(right_child.borrow().val);

                    queue.push_back(Some(right_child));
                },
                (Some(left_child), Some(right_child)) =>
                {
                    queue.push_back(Some(left_child));

                    queue.push_back(Some(right_child));
                }
                _ => ()
            }
        }

        result
    }
}